### PR TITLE
[Feature] support ignore nulls for first_value and last_value function

### LINKED ISSUE #13632: FIRST_VALUE and LAST_VALUE should support IGNORE NULLS
## Feature request

FIRST_VALUE and LAST_VALUE should support IGNORE NULLS

**Describe the solution you'd like**

Would like syntax and semantics to match vertica and oracle. Example:
LAST_VALUE(expr IGNORE NULLS)

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #13632

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by silverbullet233 — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|5|5|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|5|5|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|5|5|100.00%|[]|

File: be/src/column/column_helper.cpp [hunk 1]
```diff
@@ -200,6 +200,36 @@ ColumnPtr ColumnHelper::create_const_null_column(size_t chunk_size) {
...
+    auto& null = as_raw_column<NullableColumn>(col)->immutable_null_column_data();
+    for (size_t i = end - 1;;) {
+        if (null[i] == 0) {
+            return i;
+        }
+        if (i == start) {
+            break;
+        }
+        i--;
+    }
```
Comment (satanson) [inline be/src/column/column_helper.cpp:229]: suggest to write like this
```
    for (size_t i = end - 1; i!=start;--i) {
        if (null[i] == 0) {
            return i;
        }
    }
```

Comment (mofeiatwork) [reply be/src/column/column_helper.cpp:229]: Well, it's different:
- As `for (size_t = end-1; i != start; --i)`, the `start` is exclusive
- As `if (i == start) break; i--;`, the `start` is inclusive

File: be/src/exprs/agg/window.h [hunk 2]
```diff
@@ -519,28 +515,36 @@ class FirstValueWindowFunction<PT, Slice, StringPTGuard<PT>> final : public Wind
...
 
     void update_batch_single_state_with_frame(FunctionContext* ctx, AggDataPtr __restrict state, const Column** columns,
                                               int64_t peer_group_start, int64_t peer_group_end, int64_t frame_start,
                                               int64_t frame_end) const override {
-        if (columns[0]->is_null(frame_end - 1)) {
+        if (frame_start >= frame_end) {
             this->data(state).is_null = true;
             return;
         }
-        this->data(state).is_null = false;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply be/src/exprs/agg/window.h:536]: It is. The `is_null` is checked for several cases:
1. Empty frame: `frame_start >= frame_end`
2. Last value is null: previously `columns[0]->is_null(frame_end-1)`

For `ignore null`, there's one more cases for `is_null`:
1. Empty frame(same as before): `frame_start >= frame_end`
2. Last value is null: `columns[0]->is_null(value_index)`
3. Non-null value not found: `value_index == frame_end`

So the last two check is unfied as `        if (value_index == frame_end || columns[0]->is_null(value_index))`

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CallOperator.java [hunk 3]
```diff
@@ -48,6 +48,9 @@ public class CallOperator extends ScalarOperator {
     // The flag for distinct function
     private final boolean isDistinct;
 
+    // Ignore nulls.
+    private boolean ignoreNulls = false;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CallOperator.java:52]: Not exactly. 

The `ignoreNulls` only works for `first_value` and `last_value` function. 

It means:
- Return the first/last non-null value if exists
- Return the null if not exists

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by zuyu** [comment]
@mergify backport branch-2.5

**Comment by zuyu** [comment]
Do we have the PR to update the doc repo as well?

**Pull request merged**